Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp

```
@@ -164,4 +166,95 @@ Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailable
 }
 
+Vector<uint8_t> createIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths)
+{
+    Ref buffer = FragmentedSharedBuffer::create(data);
+    Ref imageDecoder = ImageDecoderCG::create(buffer.get(), AlphaOption::Premultiplied, GammaAndColorProfileOption::Applied);
+    imageDecoder->setData(buffer.get(), true);
+    if (imageDecoder->encodedDataStatus() == EncodedDataStatus::Error)
+        return { };
+
+    auto sourceUTI = imageDecoder->uti();
+    if (!isSupportedImageType(sourceUTI))
+        return { };
+
+    // If the image contains multiple frames, we don't know which one to use.
+    size_t frameCount = imageDecoder->frameCount();
+    if (frameCount != 1)
+        return { };
+
+    RetainPtr image = imageDecoder->createFrameImageAtIndex(0);
+    if (!image)
+        return { };
+
+    std::array<unsigned, 4> availableLengths { { 16, 32, 48, 256 } };
+    auto targetLengths = lengths.empty() ? std::span { availableLengths } : lengths;
+    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceGenericRGB));
+    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
+    String destinationUTI = "com.microsoft.ico"_s;
+    RetainPtr uti = destinationUTI.createCFString();
+    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), uti.get(), targetLengths.size(), nullptr));
+    for (auto length : targetLengths) {
+        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
+        if (!context)
+            continue;
+
+        CGContextSetInterpolationQuality(context.get(), kCGInterpolationHigh);
+        CGContextDrawImage(context.get(), CGRectMake(0, 0, length, length), image.get());
+        if (RetainPtr newImage = adoptCF(CGBitmapContextCreateImage(context.get())))
+            CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
```

Same question here. Is silently skipping one of the lengths the right thing to do?

---

Will change to early return.